// y-presence/dist/index.js
"use strict";
var c = Object.defineProperty;
var w = Object.getOwnPropertyDescriptor;
var g = Object.getOwnPropertyNames;
var m = Object.prototype.hasOwnProperty;
var A = (e, t) => {
  for (var s in t)
    c(e, s, { get: t[s], enumerable: !0 });
}, b = (e, t, s, n) => {
  if (t && typeof t == "object" || typeof t == "function")
    for (let o of g(t))
      !m.call(e, o) && o !== s && c(e, o, { get: () => t[o], enumerable: !(n = w(t, o)) || n.enumerable });
  return e;
};
var y = (e) => b(c({}, "__esModule", { value: !0 }), e);

// src/index.ts
var C = {};
A(C, {
  useSelf: () => U,
  useUsers: () => R
});
module.exports = y(C);

// src/useSelf.ts
var l = require("use-sync-external-store/shim/with-selector.js");
function U(e, t = (n) => n, s) {
  return (0, l.useSyncExternalStoreWithSelector)(
    (o) => x(e, o),
    () => i(e),
    () => i(e),
    t,
    s
  );
}
function x(e, t) {
  let s = (n, o) => {
    typeof o == "string" && o === "local" && t();
  };
  return e.on("change", s), () => e.off("change", s);
}
function i(e) {
  return e.getLocalState();
}

// src/useUsers.ts
var a = require("react"), p = require("use-sync-external-store/shim/with-selector.js");
function R(e, t = (n) => n, s) {
  let n = (0, a.useRef)();
  n.current || (n.current = new Map(e.getStates()));
  let o = (0, a.useCallback)(() => n.current ? n.current : /* @__PURE__ */ new Map(), []), u = (0, p.useSyncExternalStoreWithSelector)(
    (r) => f(e, r),
    o,
    o,
    t,
    s
  ), f = (0, a.useCallback)(
    (r, h) => {
      let S = () => {
        n.current = new Map(r.getStates()), h();
      };
      return r.on("change", S), () => r.off("change", S);
    },
    []
  );
  return u;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useSelf,
  useUsers
});
